src/Core/Managed/Shared/Extensibility/MetricAggregatorManager.cs

```
@@ -0,0 +1,295 @@
+            }
+
+            if (aggregator == null)
+            {
+                return null;
+            }
+
+            return this.aggregatorDictionary.GetOrAdd(aggregator, (aid) => { return new SimpleMetricStatisticsAggregator(); });
+        }
+
+        /// <summary>
+        /// Calculates wait time until next snapshot of the aggregators.
+        /// </summary>
+        /// <returns>Wait time.</returns>
+        private static TimeSpan GetWaitTime()
+        {
+            DateTimeOffset currentTime = DateTimeOffset.UtcNow;
+
+            double minutesFromZero = currentTime.Subtract(DateTimeOffset.MinValue).TotalMinutes;
+
+            // we want to wake up exactly at 1 second past minute
+            // to make perceived system latency look smaller
+            var nextWakeTime = DateTimeOffset.MinValue
+                .AddMinutes((long)minutesFromZero)
+                .Add(snapshotFrequency)
+                .AddSeconds(1);
+
+            TimeSpan sleepTime = nextWakeTime - DateTimeOffset.UtcNow;
+
+            // adjust wait time to a bit longer than a minute if the wake up time is within few seconds from now
+            return sleepTime < TimeSpan.FromSeconds(3) ? sleepTime.Add(snapshotFrequency) : sleepTime;
+        }
+
+        /// <summary>
+        /// Generates telemetry object based on the metric aggregator.
+        /// </summary>
+        /// <param name="aggregator">Metric aggregator.</param>
+        /// <param name="aggregatorStats">Metric aggregator statistics calculated for a period of time.</param>
+        /// <returns>Metric telemetry object resulting from aggregation.</returns>
+        private static AggregatedMetricTelemetry CreateAggergatedMetricTelemetry(MetricAggregator aggregator, SimpleMetricStatisticsAggregator aggregatorStats)
+        {
+            if ((aggregator == null) || (aggregatorStats == null) || (aggregatorStats.Count <= 0))
```

They cannot be, correct. If favor of failing fast, I'll remove the check.